[ANGLE] Fix -Wsign-compare warnings on watchOS

ddkilzer opened the pull request:
#### 076b5adb1937cf7b5f7ea2ac7b6629e0792a493b
<pre>
[ANGLE] Fix -Wsign-compare warnings on watchOS
<a href="https://bugs.webkit.org/show_bug.cgi?id=250684">https://bugs.webkit.org/show_bug.cgi?id=250684</a>
&lt;rdar://104301662&gt;

Reviewed by Kimmo Kinnunen.

On watchOS, &apos;GLintptr&apos; and &apos;ptrdiff_t&apos; are defined as signed
types, so they must be converted to an unsigned value to compare
them to std::numeric_limits&lt;uint32_t&gt;::max().

In general, this code treats signed variables as containers for
unsigned values, so checking for negative values isn&apos;t important
in this context.  Instead overflow only needs to be checked
before casting if the signed type is larger than uint32_t.

* Source/ThirdParty/ANGLE/Configurations/Base.xcconfig:
(WK_FIXME_WARNING_CFLAGS): Remove -Wno-sign-compare.

* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/FrameBufferMtl.mm:
(rx::FramebufferMtl::readPixelsToPBO const):
- Only check for overflow if the size of
  `packPixelsParams.offset` (ptrdiff_t) is larger than the size
  of uint32_t.
- Use static_cast&lt;std::make_unsigned_t&lt;&gt;&gt;() when checking for
  overflow of `packPixelsParams.offset`.

* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/VertexArrayMtl.mm:
(rx::VertexArrayMtl::convertVertexBufferGPU):
- Extract `binding.getOffset()` into a local variable.
- Only check for overflow if the size of `bindingOffset`
  (GLintptr) is larger than the size of uint32_t.
- Use static_cast&lt;std::make_unsigned_t&lt;&gt;&gt;() when checking for
  overflow of `bindingOffset`.
- Add an overflow check for `vertexConversion-&gt;offset` since
  it&apos;s a size_t value, and switch to std::min() from MIN().

Canonical link: <a href="https://commits.webkit.org/259542@main">https://commits.webkit.org/259542@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e58329e6cfbeff187aa25538789086483f1214f5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104483 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13561 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37388 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113759 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/173983 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/14666 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4481 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/96822 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112726 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110250 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11306 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94361 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/38899 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93171 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25973 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80568 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6919 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27330 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7044 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/3888 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13076 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46890 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6568 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8838 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/VertexArrayMtl.mm:
```diff
@@ -1092,7 +1092,11 @@ inline void SetDefaultVertexBufferLayout(mtl::VertexBufferLayoutDesc *layout)
+    GLintptr bindingOffset = binding.getOffset();
+
+    if constexpr (std::is_signed_v<GLintptr>)
+        ANGLE_CHECK_GL_MATH(contextMtl, bindingOffset >= 0);
```
If `GLintptr` is signed on all platforms, then we can remove the `if constexpr` line.  (But if that's true, it's odd that the sign-compare warning doesn't occur on all platforms.)

djg replied:
`GLintptr` should be signed, ptr-sized type.

kkinnunen-apple replied:
>it's odd that the sign-compare warning doesn't occur on all platforms

As far as I interpret the stackoverflow explanation:
- on 64-bit the compare is ` int64_t <= uint32_t` and type system turns it to `int64_t <= int64_t`. This is sound comparison and doesn't warrant a warning.
- on 32-bit the compare is `int32_t <= uint32_t`, which the type system turns to `uint32_t <= uint32_t`. This dangerous comparison as it inverts the result for negative numbers, and as such it should be warning. (in general, that is. For this particular case, it does not matter, as it's an no-op for 32-bit systems)

ddkilzer replied:
@kkinnunen-apple Via Slack DM you said this change was "wrong", but I'm not sure I understand "how" you think it's wrong.

Is it that this change should only exist on 32-bit systems and should not apply to 64-bit systems?

My assumption was that even if the check wasn't _strictly_ necessary for 64-bit systems, it was still a valid check since a `signed` value could still be negative, and existing code is just happily casting a signed value to an unsigned value without checking whether it's greater than or equal to zero.

kkinnunen-apple replied:
> I'm not sure I understand "how" you think it's wrong.

Just checking, did you notice the explanations above? I did try to explain the rationale.

> Is it that this change should only exist on 32-bit systems and should not apply to 64-bit systems?

No, I think most of the changes in the patches are incorrect "in spirit" on 32-bit systems. (And unneeded on 64-bit systems) 

The logic is: just because the intptr type is signed, it does not mean that it should be compared < 0 before converting to a 32-bit value for an algorithm that has 32 bit limitation.

So as an example: the binding offset here comes from the GL API, `pointer` argument:
`void VertexAttribPointer( uint index, int size, enum type, boolean normalized, sizei stride, const void *pointer );`

If the caller wants to specify 2GB vertex attrib offset, they say `glVertexAttribPointer(..., reinterpret_cast<void*>(0x80000000));`.

Now, suppose the previous code was:
```
uint32_t offset  = static_cast<uint32_t>( binding.getOffset());
```

So you cannot just blanket change that to:
```
ANGLE_CHECK_GL_MATH(contextMtl, binding.getOffset() >= 0);
uint32_t offset  = static_cast<uint32_t>( binding.getOffset());
```

If you do this, you might break some case that was theoretically working before.

This is because it's not a signed value to begin with -- it is just stored in a signed type for various convoluted reasons.

So to fix the signedness warnings you need to fix them in a way that preserves the semantics of the value being used.

So in this binding case:
```
ANGLE_CHECK_GL_MATH(contextMtl, binding.getOffset() <= std::numeric_limits<uint32_t>::max());
```

This is a valid check, as it appears that the algorithm is limited to 32 bits.

It's a pointless check on 32-bit platforms, as the GLintptr_t, int, is promoted to unsigned and <= max unsigned, which is a tautology.

It's triggering a signedness comparison warning, as it is promoting int to unsigned.

The warning should be fixed in a way that preserves the existing algorithm, unless of course the existing algorithm is broken.

ddkilzer replied:
Thanks for the explanation!  I think I understand now.

djg reviewed (approved):
Looks OK to me, but I feel that we should pass `GLintptr` through to the compute shader instead of converting to `uint32_t`, but that can we for another patch.

kkinnunen-apple commented on Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/FrameBufferMtl.mm:
```diff
@@ -1655,9 +1655,11 @@ void RoundValueAndAdjustCorrespondingValue(float a,
+    ANGLE_MTL_CHECK(contextMtl, packPixelsParamsOffset >= 0, GL_INVALID_OPERATION);
+    ANGLE_MTL_CHECK(contextMtl, static_cast<std::make_unsigned_t<ptrdiff_t>>(packPixelsParamsOffset) <= std::numeric_limits<uint32_t>::max(),
                     GL_INVALID_OPERATION);
-    uint32_t offset = static_cast<uint32_t>(packPixelsParams.offset);
```
If I understand correctly, I think this here is more closer to the GL purpose.

The idea of the GL API, and the static cast as it is written here: 
1) if you want to upload 2gbs of data on 32-bit platform, you can (e.g. GLintptr is used to just convey the bits).  
2) if you want to upload 30gbs of data on 64-bit platform, you can (e.g. API uses GLintptr which is 64-bit)

GLintptr is not used so that the driver would compare > 0.

ddkilzer replied:
> GLintptr is not used so that the driver would compare > 0.

I'm assuming that "the driver" in this context would mean "GPU driver code would compare > 0, but not the code in this file".

kkinnunen-apple commented on Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/FrameBufferMtl.mm:
```diff
@@ -1655,9 +1655,11 @@ void RoundValueAndAdjustCorrespondingValue(float a,
 
-    ANGLE_MTL_CHECK(contextMtl, packPixelsParams.offset <= std::numeric_limits<uint32_t>::max(),
+    ptrdiff_t packPixelsParamsOffset = packPixelsParams.offset;
+    ANGLE_MTL_CHECK(contextMtl, packPixelsParamsOffset >= 0, GL_INVALID_OPERATION);
```
I don't think this is correct.. IIUC the intention of the GL API, the GLintptr is not used in the APIs for the sake of the driver comparing signedness.

It's used to smuggle 32-bit data on 32-bit platforms and 64-bit data on 64-bit platforms. The data is interpreted case-by case.

I believe you want, in spirit,
```
if constexpr(sizeof(packPixelsParams.offset) > sizeof(uint32))
  ANGLE_MTL_CHECK(contextMtl, packPixelsParams.offset <= std::numeric_limits<uint32_t>::max(), GL_INVALID_OPERATION);
uint32_t offset = static_cast<uint32_t>(packPixelsParams.offset);
```

Not sure if it will work as-is on watchOS.

ddkilzer replied:
> It's used to smuggle 32-bit data on 32-bit platforms and 64-bit data on 64-bit platforms. The data is interpreted case-by case.

Okay, I hadn't anticipated that the variable would be used as a generic container—I figured an overflow would be meaningful.

That's where my assumptions differed from yours.  Thanks!

kkinnunen-apple commented on Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/VertexArrayMtl.mm:
```diff
@@ -1101,7 +1105,7 @@ inline void SetDefaultVertexBufferLayout(mtl::VertexBufferLayoutDesc *layout)
         static_cast<VertexConversionBufferMtl *>(conversion);
     params.srcBuffer            = srcBuffer->getCurrentBuffer();
     params.srcBufferStartOffset = static_cast<uint32_t>(
-        MIN(static_cast<GLintptr>(vertexConversion->offset), binding.getOffset()));
```
I think this existing `min` operation is dubious.
E.g. if the intention is to interpret the offset as uint32_t, then the min should operate on that, not that it first operates on int32_t and then converts to its inverse. 

Of course depends on whether the algorithm using srBufferStartOffset takes it into account that the srcBufferStartOffset is uint32_t but expects the 2s complement minus value to wrap around correctly and actually decrease the value it is being added to..

ddkilzer replied:
Got it.  Looks to me like `srcBufferStartOffset` is used as a `uint32_t` everywhere.

Commit:
[ANGLE] Fix -Wsign-compare warnings on watchOS
https://bugs.webkit.org/show_bug.cgi?id=250684
<rdar://104301662>

Reviewed by Kimmo Kinnunen.

On watchOS, 'GLintptr' and 'ptrdiff_t' are defined as signed
types, so they must be converted to an unsigned value to compare
them to std::numeric_limits<uint32_t>::max().

In general, this code treats signed variables as containers for
unsigned values, so checking for negative values isn't important
in this context.  Instead overflow only needs to be checked
before casting if the signed type is larger than uint32_t.

* Source/ThirdParty/ANGLE/Configurations/Base.xcconfig:
(WK_FIXME_WARNING_CFLAGS): Remove -Wno-sign-compare.

* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/FrameBufferMtl.mm:
(rx::FramebufferMtl::readPixelsToPBO const):
- Only check for overflow if the size of
  `packPixelsParams.offset` (ptrdiff_t) is larger than the size
  of uint32_t.
- Use static_cast<std::make_unsigned_t<>>() when checking for
  overflow of `packPixelsParams.offset`.

* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/VertexArrayMtl.mm:
(rx::VertexArrayMtl::convertVertexBufferGPU):
- Extract `binding.getOffset()` into a local variable.
- Only check for overflow if the size of `bindingOffset`
  (GLintptr) is larger than the size of uint32_t.
- Use static_cast<std::make_unsigned_t<>>() when checking for
  overflow of `bindingOffset`.
- Add an overflow check for `vertexConversion->offset` since
  it's a size_t value, and switch to std::min() from MIN().

Canonical link: https://commits.webkit.org/259542@main
